src/concepts/lazy.md

```diff
@@ -0,0 +1,177 @@
+The notion of _need_ will be subjective relative to what's being implemented.
+
+### Introducing Laziness with Thunks
+
+SML is an eager language. But what if we wanted to simulate lazy evaluation?
+First we need a way to wait until we _need_ an argument to evaluate it. This
+is where **thunks** come into play. A **thunk** is a function of type
+`unit -> 'a`. They help us represent this notion of "evaluation by need" by
+wrapping an expression in a function, which delays the computation of that expression.
+
+Let's say we have some function `f` and argument `x`. When we run `f x`, `x`
+will get evaluated immediately. However, by wrapping `x` inside a function,
+we can delay its computation _(assume `f`'s type is correct)_: `f (fn () => x)`.
+Since **_functions are values_**, this lambda function is already fully evaluated.
+However, `x` has yet to be evaluated! Now we can wait to evaluate `x` by passing
+in `() : unit` to the lambda function at some point.
+
+Let's see how that changes things in our earlier example. Here, we change `square`
+into a lazy version that accepts a **thunk** instead of an `int`.
+
+```sml
+val double : int -> int = fn n => 2 * n
+fun square (f : unit -> int) : int = f () * f ()
+
+(* Lazy *)
+    double square (fn () => double 2)
+==> double ((fn () => double 2) () * (fn () => double 2) ())
+==> double (double 2 * double 2)
+==> double (4 * 4)
+==> double 16
+==> 32
+```
+
+You can see that with lazy evaluation, we put off the computation of `double 2`
+until we "needed it". That "need" is subjective, but the key idea is that we
+delayed the computation of `double 2` by wrapping it a **thunk**. Lets see
+what this looks like in normal algebra.
+
+```sml
+(* Eager *)                  (* Lazy *)
+    2 * (2 * 2)^2                2 * (2 * 2)^2
+==> 2 * (4)^2                ==> 2 * ((2 * 2) * (2 * 2))
+==> 2 * (4 * 4)              ==> 2 * (4 * 4)
+==> 2 * 16                   ==> 2 * 16
+==> 32                       ==> 32
+```
+
+Since `square` is now the lazy function, we are delaying the evaluation of
+`square`'s arguments. That's why in eager evaluation, we just do `2 * 2 ==> 4`
+before we square that value, whereas in lazy evaluation, we square the
+unevaluated arguments by doing `(2 * 2) * (2 * 2)`.
+
+_Note, we let the type of a **thunk** be `unit -> 'a` just because it's convenient._
+_Theoretically I could use a different type to represent thunks since all we need_
+_is a way to wrap an expression in a function to delay its computation._
+
+## Lazy Lists
+
+Now that we have established this notion of **_laziness_** in SML, we can do
+even fancier things, like create infinite data structures! Let's first look at
+a lazy list and compare it to a normal list.
+
+```sml
+datatype 'a list     = Nil | Cons of 'a * 'a list
+datatype 'a lazylist =       Cons of 'a * (unit -> 'a lazylist)
+```
+
+You'll see that a `lazylist` is very similar to normal `list`s. The only difference
```

Comment:
If you define a "`lazylist`" to be only infinite then I don't think they're very similar at all! And regardless of what we do on a lab I definitely wouldn't call this datatype a lazy list (probably an infinite stream instead).